File: Java/majority.java
//Given an array of size n, find the majority element. The majority element is the element that appears more than floor(n/2) times.
//
//You may assume that the array is non-empty and the majority element always exist in the array.
//
//Example :
//
//Input : [2, 1, 2]
//Return  : 2 which occurs 2 times which is greater than 3/2. 
//

public class Solution {
	public int majorityElement(final List<Integer> A) {
	    
	    if (A == null)
	        return -1;
	        
	    int maj = A.get(0);
	    int count = 1;
	    int n = A.size();
	    
	    for (int i = 1; i < n; i++) {
	        if (A.get(i) == maj) {
	            count++;
	        } else if (count == 1) {
	            maj = A.get(i);
	        } else {
	            count--;
	        }
	    }
	    
	    count = 0;
	    
	    for (int i = 0; i < n; i++) {
	        if (A.get(i) == maj)
	            count++;
	    }
	    
	    if (count > n / 2)
	        return maj;
	        
	    return -1;
	    
	    
	}
}

File: JavaScript/majority.js
/*

Given an array of length n, find an element that occurs more than n / 2 times

Input : {3, 3, 4, 2, 4, 4, 2, 4, 4}
Output : 4 

Input : {3, 3, 4, 2, 4, 4, 2, 4}
Output : -1

*/

const getMajority = arr => {

	// map holds key value pairs, the value for each key is the frequency of occurence
	let count = new Map();
	let n = arr.length;

	// iterating through the array
	for(let x of arr) {

		// if the element has not been seen before, set its occurences to 1
		if(!count.has(x)) count.set(x, 1);

		// otherwise increment
		else count.set(x, count.get(x) + 1);

		// if it becomes a majority element return it
		if(count.get(x) > n / 2) return x;
	}

	// if it reaches here then a majority element was not found and hence return -1
	return -1;
}

export { getMajority };